src/librustc_resolve/lib.rs

```diff
@@ -4449,7 +4451,12 @@ fn show_candidates(err: &mut DiagnosticBuilder,
             } else {
                 "\n"
             };
-            *candidate = format!("use {};\n{}", candidate, additional_newline);
+            let crate_prefix = if crate_in_paths {
```

Comment:
Hmm, now i'm worried that this will apply even when it shouldn't. Consider this testcase:

```rust
fn main() {
    let x: &Debug = &22;
}
```

This currently prints:

```
error[E0412]: cannot find type `Debug` in this scope
 --> src/main.rs:2:13
  |
2 |     let x: &Debug = &22;
  |             ^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
  |
1 | use std::fmt::Debug;
  |
```

and I think we do *not* want it to print `use crate::std::fmt::Debug`.

Reply:
It looks like these are produced by this function:
https://github.com/rust-lang/rust/blob/a32e979ba53af4d4654bfad7babdb9f8c197c3a2/src/librustc_resolve/lib.rs#L4003-L4015

I suspect we want to extend this struct

https://github.com/rust-lang/rust/blob/a32e979ba53af4d4654bfad7babdb9f8c197c3a2/src/librustc_resolve/lib.rs#L87-L90

with an additional field, indicating whether the first link in the path is an `extern crate` or not.  Notice that the function has the ability to check whether it is an `extern crate`:

https://github.com/rust-lang/rust/blob/a32e979ba53af4d4654bfad7babdb9f8c197c3a2/src/librustc_resolve/lib.rs#L4030-L4031

In fact, we are already threading this `in_module_is_extern` state around -- that is *almost* what we want -- it tells us if the path contains an `extern crate` *anywhere*. But we just want to know if the path contains an `extern crate` *in the first link*.

Actually, reading this code, I think that likely these suggestions will break once we start removing `extern crate` entries. It may be that the better approach is to fix that problem, basically by iterating over the extern crates that were given on the command line first. The advantage of that is that we would not need any special logic for `extern crate`. So yeah scratch what I wrote above.

Reply:
I [left some notes in the issue](https://github.com/rust-lang/rust/issues/51212#issuecomment-396190894) laying out a possible alternative plan.